network/common/src/main/java/org/apache/spark/network/protocol/MessageEncoder.java

```diff
@@ -72,9 +80,84 @@ public void encode(ChannelHandlerContext ctx, Message in, List<Object> out) {
     in.encode(header);
     assert header.writableBytes() == 0;
 
-    out.add(header);
     if (body != null && bodyLength > 0) {
-      out.add(body);
+      out.add(new MessageWithHeader(header, headerLength, body, bodyLength));
+    } else {
+      out.add(header);
     }
   }
+
+  /**
+   * A wrapper message that holds two separate pieces (a header and a body) to avoid
+   * copying the body's content.
+   */
+  private static class MessageWithHeader extends AbstractReferenceCounted implements FileRegion {
```

Comment:
I will take care to add support in netty



> Am 31.03.2015 um 05:46 schrieb Reynold Xin <notifications@github.com>:
> 
> In network/common/src/main/java/org/apache/spark/network/protocol/MessageEncoder.java:
> 
> >      }
> >    }
> > +
> > +  /**
> > +   * A wrapper message that holds two separate pieces (a header and a body) to avoid
> > +   * copying the body's content.
> > +   */
> > +  private static class MessageWithHeader extends AbstractReferenceCounted implements FileRegion {
> Yes - I think epoll won't work and we should just disable it. We'd need more work on Netty proper to support epoll for this.
> 
> —
> Reply to this email directly or view it on GitHub.
>